Describe key

andyjda opened the pull request:
See [describe-key-README](https://github.com/andyjda/god-mode/blob/describe-key-readme/describe-key-README.org) for a detailed overview of approach and trade-offs.

All the relevant code should be at the bottom of `god-mode.el`.
Be sure to check the portion of the README that I added.

Let me know if you have any questions or if you want me to make any changes

darth10 commented on god-mode.el:
```diff
@@ -461,6 +462,78 @@ be ignored by `god-execute-with-current-bindings'."
+	(if command
+	    (progn
+	      (add-hook 'help-fns-describe-function-functions
+			#'god-mode--help-fn-describe-function)
```
Since you're using hooks, is it possible to add the hook when `god-local-mode` is enabled, and remove it when the mode is disabled?

andyjda replied:
I don't really think this is doable without adding a lot of complexity.
This because of two reasons:

1. One problem with adding advices when `god-local-mode` is enabled, and removing them when it is disabled, is that we may have different "instances" of `god-local-mode` across different buffers. This creates a problem if, for example, we have a workflow along these lines:

- enable `god-local-mode` in buffer A. The hooks/advices for `god-mode-describe-key` are added
- switch to another buffer B, enable `god-local-mode`. The hooks/advices are "added" but since they are already there, this makes no difference
- disable `god-local-mode` in buffer B. The hooks are removed
- switch back to buffer A. Now the hooks/advices necessary for `god-mode-describe-key` have been removed, and the function doesn't work properly anymore.

I can't think of simple ways to fix this.

2. Another problem with this specific function, is that it is added to the describe _function_ hook, so if we simply add it as a hook when we enable `god-mode,` any call to `describe-function` would mistakenly try to lookup the inputted keys and their god-mode translation. The way to fix this could be to make `god-mode--help-fn-describe-function` discriminate whether it is being called from `god-mode-describe-key` or from a "regular" call to `describe-function`, and not do anything in the latter case. But I can't think of a way to do this without introducing a new variable: `god-mode-describe-key` could set a boolean to True when called and then set it back to False after the call to `describe-function`. To me this approach seems more complex and unnecessarily clunky than the current one.

darth10 replied:
Good point. Let's leave this as is.

darth10 commented on god-mode.el:
```diff
@@ -461,6 +462,78 @@ be ignored by `god-execute-with-current-bindings'."
+  (if (not god-translate-key-for-description)
+      (call-interactively #'describe-key)
+    (progn
+      (message "Describe the following god-mode key: ")
```
Is this `message` form needed?

andyjda replied:
It doesn't have to say exactly this, but I do think we want to prompt the user for a key-sequence. 
It's good to show this message so the user knows that Emacs is waiting on input from them. I chose the sentence "Describe the following god-mode key:" because I wanted it to mirror the sentence used by the regular `describe-key`, while alerting the user that `god-mode-describe-key` works best for keys that are specific to `god-mode`, and not for clicks and menu items (unlike `describe-key`).

darth10 commented on god-mode.el:
```diff
@@ -461,6 +462,78 @@ be ignored by `god-execute-with-current-bindings'."
+      (call-interactively #'describe-key)
+    (progn
+      (message "Describe the following god-mode key: ")
+      (advice-add #'god-mode-lookup-command :filter-args
```
Similar to my other comment related to hooks for `describe-function`, can this advice function be added/removed depending on whether `god-local-mode` is enabled/disabled?

andyjda replied:
The same problems apply here:

1. If we remove the advice when `god-local-mode` is disabled, the advice would be removed for all buffers, even ones that might still have `god-local-mode` enabled
2. This advice doesn't apply to `describe-function`, but it does apply to `god-mode-lookup-command`. Do we want to keep track of the "god-mode key-sequence" for every command? To me it seems this is useless outside of `god-mode-describe-key`, so I assumed a one-time advice is preferable.

darth10 commented:
Also, this branch needs to be rebased on master, as there are commits from previous PRs showing up.

darth10 commented on god-mode.el:
```diff
@@ -461,6 +462,78 @@ be ignored by `god-execute-with-current-bindings'."
                        universal-argument
                        universal-argument-more)))
 
+(defcustom god-translate-key-for-description
```
I think this custom var is redundant and should be removed, as `god-mode-describe-key` is an entirely separate function.
If someone doesn't want to use this feature, they would simply not use or bind the `god-mode-describe-key` function.

andyjda replied:
Good point. One reason I wanted to keep it around is because `god-mode-describe-key` is a bit limited compared to the regular `describe-key` (can't describe mouse clicks, menu items, or key bindings outside of `god-mode`). But I think the fix for this is simply to specify it in the README: users should know that if they want to describe (eg) a menu item,  they can simply turn off `god-mode` temporarily.

I will fix this in the next merge

darth10 commented on god-mode.el:
```diff
@@ -189,7 +190,7 @@ If it was not active when `god-local-mode-pause' was called, nothing happens."
 
 (define-key universal-argument-map (kbd "u")
-  #'god-mode-maybe-universal-argument-more)
+	    #'god-mode-maybe-universal-argument-more)
```
This indentation seems off. Can it be reverted?

andyjda replied:
It was the result of applying `indent-region`. But I can revert it back to what it was before

darth10 commented:
Also, this branch needs to be rebased on master, as there are commits from previous PRs showing up.

andyjda commented:
- rebased. conflicts should be gone
 - removed custom variable `god-translate-key-for-description`
 - reverted indentation

See my comments about the other proposed changes. If you can think of a good way to deal with the hooks/functions, let me know. I am not too familiar with their underlying implementation: to me it looks like the current approach is the simplest, but if there's a better one I am down to work on it

darth10 commented on god-mode.el:
```diff
@@ -463,6 +464,70 @@ be ignored by `god-execute-with-current-bindings'."
+  "Describe a key-sequence as interpreted by `god-mode'.
+Prompt for a key sequence, use `god-mode-lookup-key-sequence' to translate it
+into the appropriate command, and use `describe-function' to describe it.
+Only applied when `god-translate-key-for-description' is t:
+when nil, `describe-key' is called instead"
```
This part of the docstring that refers to `god-translate-key-for-description` is redundant now.

darth10 commented on .github/README.md:
```diff
@@ -150,6 +150,12 @@ This package defines the following key mappings:
+You can use `god-mode-describe-key` to check what command would be triggered by particular keys. This function works similarly to `describe-key`: it prompts for a key combination, translates it into a command, and display its information in the Help buffer.
+Note that `god-mode-describe-key` is only able to interpret key-bindings that are specific to `god-mode`. For other key-bindings, mouse-clicks, and menu items, it's better to use `describe-key`.
+By default, `god-mode-describe-key` is bound to `C-h k` in `god-local-mode-map`.
+You can turn off this feature by setting the custom variable `god-translate-key-for-description` to `nil`.
```
This line that talks about `god-translate-key-for-description` is redundant and should be removed.

darth10 reviewed (commented):
Thanks for putting in those changes @andyjda!
I've tested out the `god-mode-describe-key` function and it's looking good!

I've left a few more comments around supporting older Emacs versions (< 28.1).

andyjda commented on god-mode.el:
```diff
@@ -463,6 +464,70 @@ be ignored by `god-execute-with-current-bindings'."
+But in our case it's redundant"
+  (insert
+   (format-message "\n  %s\n  %s%s\n  %s%s%s\n\n"
+		   "(`god-local-mode' is enabled. "
```
Didn't know about the style, guide, thanks for letting me know!

andyjda commented:
I included all your proposed changes. 
Let me know if there's anything else you'd like me to change. Thanks :)

darth10 reviewed (approved):
Looks good! Thanks for the contribution again @andyjda.

darth10 commented:
Closes #144.
